### ISSUE TITLE
Problem with ScrollViewer event ScrollChanged

### ISSUE THREAD

**Issue by phanthanhdat077** [dotnet/wpf#9194]
I'm learning WPF and I have a problem with ScrollViewer when I scroll to VerticalOffset = 0 and add more items but when it's Insert new item it can't stop scrolling and add items forever. Someone help me check.
And how to preserve a bound scroll position when the collection changes. 

Before scroll:
![image](https://github.com/dotnet/wpf/assets/34691283/d14a4606-5aed-41c9-9ff1-adad2c35eb7d)

After scrolling to VerticalOffset 0
![image](https://github.com/dotnet/wpf/assets/34691283/fac537c8-1ac6-4463-a425-c7264cafdad8)

[WpfApp8.zip](https://github.com/user-attachments/files/15530594/WpfApp8.zip)

**Comment by huiliuss** [comment]
Problem causes：
In the method bound to LoadMoreCommand, p.ScrollToVerticalOffset(p.ScrollableHeight / 2); This code will not be executed immediately, and it will be reflected in the application after the CanExecute judgment is performed. Therefore, p.VerticalOffset == 0 is always judged as true. GetData(20); is always executed.

Solution:
Wait for p.ScrollToVerticalOffset(p.ScrollableHeight / 2); Apply the item exactly and then execute GetData(20);

 

Add a field in MainViewModel

```C#

private bool CanAddData = false;

```

 

Adjust the code

```C#

                if(CanAddData)

                {

                    GetData(20);

                    CanAddData = false;

                }

 

                if (p == null || p.VerticalOffset == 0 || IsTheEndChat || p.ExtentHeight == 0)

                {

                    IsLoad = true;

                    Scroll = p;

                    p.ScrollToVerticalOffset(p.ScrollableHeight / 2);

                    CanAddData = true;

                    //GetData(20);    

                    IsTheEndChat = ListChat.Count == 0;

                    IsLoad = false;

                }

```

**Comment by phanthanhdat077** [comment]
> Problem causes： In the method bound to LoadMoreCommand, p.ScrollToVerticalOffset(p.ScrollableHeight / 2); This code will not be executed immediately, and it will be reflected in the application after the CanExecute judgment is performed. Therefore, p.VerticalOffset == 0 is always judged as true. GetData(20); is always executed.
> 
> Solution: Wait for p.ScrollToVerticalOffset(p.ScrollableHeight / 2); Apply the item exactly and then execute GetData(20);
> 
> Add a field in MainViewModel
> 
> ```cs
> private bool CanAddData = false;
> ```
> 
> Adjust the code
> 
> ```cs
>                 if(CanAddData)
> 
>                 {
> 
>                     GetData(20);
> 
>                     CanAddData = false;
> 
>                 }
> 
>  
> 
>                 if (p == null || p.VerticalOffset == 0 || IsTheEndChat || p.ExtentHeight == 0)
> 
>                 {
> 
>                     IsLoad = true;
> 
>                     Scroll = p;
> 
>                     p.ScrollToVerticalOffset(p.ScrollableHeight / 2);
> 
>                     CanAddData = true;
> 
>                     //GetData(20);    
> 
>                     IsTheEndChat = ListChat.Count == 0;
> 
>                     IsLoad = false;
> 
>                 }
> ```

Can you tell me if there is any way to scroll to ExtentHeightChange after calling GetData? I tried to calculate with ExtentHeight but failed can't preserve a bound scroll position

**Comment by hongruiyu** [comment]
ExtentHeight is the value of the vertical size of the extended data block. You can think of it as the size that can be slid. The more data there is, the larger the size that can be slid. In order to wrap more data, ExtentHeight cannot be called back.
 
But when GetData(20); is executed, ExtentHeight will not be modified immediately. During this period, you can do what you want.

**Comment by phanthanhdat077** [comment]
> ExtentHeight is the value of the vertical size of the extended data block. You can think of it as the size that can be slid. The more data there is, the larger the size that can be slid. In order to wrap more data, ExtentHeight cannot be called back.
> 
> But when GetData(20); is executed, ExtentHeight will not be modified immediately. During this period, you can do what you want.

yes after I tried scrolling to the last ExtentHeight before GetData but after GetData it's changed again. Can you tell me any ideas that can help?

**Comment by hongruiyu** [comment]
The simplest way:
You can take the space occupied by GetData(20) into account and set the position you need to slide to before GetData(20).
 
Add a few fields
```C#
private bool InitBoxHeight = false;
private double BoxHeight;
private bool CanAddData = false;
```
 
Use p.ScrollToVerticalOffset(BoxHeight); to set the position to slide in advance
```C#
if (CanAddData)
{
 
    GetData(20);
    CanAddData = false;
}
if ((p == null || p.VerticalOffset == 0 || IsTheEndChat || p.ExtentHeight == 0))
{
    IsLoad = true;
    Scroll = p;
    if(!InitBoxHeight)
    {
        BoxHeight = p.ExtentHeight*2;
        InitBoxHeight = true;
    }
    p.ScrollToVerticalOffset(BoxHeight);
    CanAddData = true;               
    IsTheEndChat = ListChat.Count == 0;
    IsLoad = false;
}
```
